File: c_impl/src/dtd.c
/* dtd.c — port of aec.py DtdEstimator. */
#include "dtd.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>

static void common_init(DtdEstimator* d, double attack, double release,
                        int warmup_frames) {
    d->confidence = 0.0;
    d->attack     = attack;
    d->release    = release;
    d->warmup_frames = warmup_frames;
    d->frame_count   = 0;
    d->far_abs_buffer = NULL; d->buf_len = 0; d->buf_idx = 0;
    d->geigel_threshold = 0.5; d->hangover_max = 0; d->hangover_count = 0;
    d->divergence_factor = 1.5;
    d->n_freqs = 0;
    d->S_ex = NULL; d->S_ee = NULL; d->S_xx = NULL; d->voice_weight = NULL;
    d->coh_alpha = 0.85; d->coh_high = 0.6; d->coh_low = 0.3;
    d->coh_energy_floor = 0.01; d->coh_abs_floor = 1e-6;
}

void dtd_init_geigel(DtdEstimator* d, int window_blocks,
                        double geigel_threshold, int hangover_max,
                        double attack, double release, int warmup_frames) {
    common_init(d, attack, release, warmup_frames);
    d->mode = DTD_GEIGEL;
    int wb = window_blocks > 0 ? window_blocks : 1;
    d->far_abs_buffer = (double*)calloc((size_t)wb, sizeof(double));
    d->buf_len = wb;
    d->geigel_threshold = geigel_threshold;
    d->hangover_max = hangover_max;
}

void dtd_init_divergence(DtdEstimator* d, double divergence_factor,
                            double attack, double release, int warmup_frames) {
    common_init(d, attack, release, warmup_frames);
    d->mode = DTD_DIVERGENCE;
    d->divergence_factor = divergence_factor;
}

void dtd_init_coherence(DtdEstimator* d, int n_freqs,
                           int sample_rate, int block_size,
                           double coh_alpha, double coh_high, double coh_low,
                           double coh_energy_floor, double coh_abs_floor,
                           double attack, double release, int warmup_frames) {
    common_init(d, attack, release, warmup_frames);
    d->mode = DTD_COHERENCE;
    d->n_freqs = n_freqs;
    d->coh_alpha = coh_alpha;
    d->coh_high  = coh_high;
    d->coh_low   = coh_low;
    d->coh_energy_floor = coh_energy_floor;
    d->coh_abs_floor    = coh_abs_floor;
    d->S_ex = (Complex*)calloc((size_t)n_freqs, sizeof(Complex));
    d->S_ee = (float*)calloc((size_t)n_freqs, sizeof(float));
    d->S_xx = (float*)calloc((size_t)n_freqs, sizeof(float));
    d->voice_weight = (float*)malloc((size_t)n_freqs * sizeof(float));
    double freq_per_bin = (double)sample_rate / (double)block_size;
    for (int k = 0; k < n_freqs; ++k) {
        double f = k * freq_per_bin;
        if (f >= 300.0 && f <= 4000.0) d->voice_weight[k] = 3.0f;
        else if (f < 100.0 || f > 6000.0) d->voice_weight[k] = 0.3f;
        else d->voice_weight[k] = 1.0f;
    }
}

void dtd_free(DtdEstimator* d) {
    free(d->far_abs_buffer);
    free(d->S_ex); free(d->S_ee); free(d->S_xx); free(d->voice_weight);
    d->far_abs_buffer = NULL;
    d->S_ex = NULL; d->S_ee = NULL; d->S_xx = NULL; d->voice_weight = NULL;
}

void dtd_reset(DtdEstimator* d) {
    d->confidence = 0.0;
    d->frame_count = 0;
    d->buf_idx = 0;
    d->hangover_count = 0;
    if (d->far_abs_buffer) memset(d->far_abs_buffer, 0,
                                  (size_t)d->buf_len * sizeof(double));
    if (d->S_ex) memset(d->S_ex, 0, (size_t)d->n_freqs * sizeof(Complex));
    if (d->S_ee) memset(d->S_ee, 0, (size_t)d->n_freqs * sizeof(float));
    if (d->S_xx) memset(d->S_xx, 0, (size_t)d->n_freqs * sizeof(float));
}

static void update_confidence(DtdEstimator* d, int detected) {
    if (detected) {
        d->hangover_count = d->hangover_max;
        d->confidence += d->attack;
        if (d->confidence > 1.0) d->confidence = 1.0;
    } else if (d->hangover_count > 0) {
        d->hangover_count--;
        d->confidence -= d->release * 0.5;
        if (d->confidence < 0.0) d->confidence = 0.0;
    } else {
        d->confidence -= d->release;
        if (d->confidence < 0.0) d->confidence = 0.0;
    }
}

static void detect_geigel(DtdEstimator* d, const float* near_end, int n,
                          const float* far_end, int nf) {
    double far_abs_max = 0.0;
    for (int i = 0; i < nf; ++i) {
        double a = fabs((double)far_end[i]);
        if (a > far_abs_max) far_abs_max = a;
    }
    d->far_abs_buffer[d->buf_idx] = far_abs_max;
    d->buf_idx = (d->buf_idx + 1) % d->buf_len;
    double far_max = 0.0;
    for (int i = 0; i < d->buf_len; ++i) {
        if (d->far_abs_buffer[i] > far_max) far_max = d->far_abs_buffer[i];
    }
    double near_max = 0.0;
    for (int i = 0; i < n; ++i) {
        double a = fabs((double)near_end[i]);
        if (a > near_max) near_max = a;
    }
    int detected = (far_max > 1e-6) && (near_max > d->geigel_threshold * far_max);
    update_confidence(d, detected);
}

static void detect_divergence(DtdEstimator* d, const float* near_end, int n,
                              const float* output) {
    double out_e = 0.0, near_e = 0.0;
    double out_p = 0.0, near_p = 0.0;
    for (int i = 0; i < n; ++i) {
        double on = output[i], ne = near_end[i];
        out_e += on * on; near_e += ne * ne;
        double ao = fabs(on), an = fabs(ne);
        if (ao > out_p) out_p = ao;
        if (an > near_p) near_p = an;
    }
    out_e /= n; near_e /= n;
    if (near_e < 1e-10 && near_p < 1e-6) {
        d->confidence -= d->release;
        if (d->confidence < 0.0) d->confidence = 0.0;
        return;
    }
    double energy_ratio = (near_e > 1e-10) ? out_e / (near_e + 1e-10) : 0.0;
    double peak_ratio   = (near_p > 1e-6) ? out_p / (near_p + 1e-10) : 0.0;
    double ratio = energy_ratio > peak_ratio ? energy_ratio : peak_ratio;
    const double mild = 1.2;
    if (ratio > d->divergence_factor) {
        d->confidence += d->attack;
        if (d->confidence > 1.0) d->confidence = 1.0;
    } else if (ratio > mild) {
        d->confidence += d->attack * (ratio - mild);
        if (d->confidence > 1.0) d->confidence = 1.0;
    } else {
        double scale = 1.0 - ratio;
        if (scale < 0.2) scale = 0.2;
        d->confidence -= d->release * (1.0 + 4.0 * scale);
        if (d->confidence < 0.0) d->confidence = 0.0;
    }
}

static void detect_coherence(DtdEstimator* d,
                             const Complex* error_spec, const Complex* far_spec) {
    double a = d->coh_alpha, oa = 1.0 - d->coh_alpha;
    /* cross = error * conj(far) */
    for (int k = 0; k < d->n_freqs; ++k) {
        float er = error_spec[k].r, ei = error_spec[k].i;
        float fr = far_spec[k].r,   fi = far_spec[k].i;
        float cr = er * fr + ei * fi;
        float ci = ei * fr - er * fi;
        d->S_ex[k].r = (float)(a * d->S_ex[k].r + oa * cr);
        d->S_ex[k].i = (float)(a * d->S_ex[k].i + oa * ci);
        float epsd = er * er + ei * ei;
        float xpsd = fr * fr + fi * fi;
        d->S_ee[k] = (float)(a * d->S_ee[k] + oa * epsd);
        d->S_xx[k] = (float)(a * d->S_xx[k] + oa * xpsd);
    }
    double num = 0.0, den = 0.0;
    double sum_ee = 0.0, sum_xx = 0.0;
    for (int k = 0; k < d->n_freqs; ++k) {
        double w = d->voice_weight[k];
        double sx_r = d->S_ex[k].r, sx_i = d->S_ex[k].i;
        num += w * (sx_r * sx_r + sx_i * sx_i);
        den += w * (double)d->S_ee[k] * (double)d->S_xx[k];
        sum_ee += d->S_ee[k];
        sum_xx += d->S_xx[k];
    }
    double coherence = num / (den + 1e-10);
    int has_energy = (sum_ee > d->coh_energy_floor * sum_xx)
                  && (sum_xx > 1e-10) && (sum_ee > d->coh_abs_floor);
    if (coherence > d->coh_high) {
        update_confidence(d, 0);
    } else if (coherence < d->coh_low && has_energy) {
        update_confidence(d, 1);
    } else {
        d->confidence -= d->release * 0.5;
        if (d->confidence < 0.0) d->confidence = 0.0;
    }
}

double dtd_detect_block(DtdEstimator* d,
                           const float* near_end, int near_len,
                           const float* far_end, int far_len,
                           const float* output,
                           const Complex* error_spec,
                           const Complex* far_spec) {
    d->frame_count++;
    if (d->frame_count < d->warmup_frames) return 0.0;
    switch (d->mode) {
        case DTD_GEIGEL:
            detect_geigel(d, near_end, near_len, far_end, far_len);
            break;
        case DTD_COHERENCE:
            if (error_spec && far_spec) detect_coherence(d, error_spec, far_spec);
            break;
        case DTD_DIVERGENCE:
            if (output) detect_divergence(d, near_end, near_len, output);
            break;
    }
    return d->confidence;
}

File: python/modules/dtd.py
"""DtdEstimator — Double-Talk Detector with per-mode strategy.

Extracted from ``aec.py`` during refactor R.6. Has its own module
(separate from ``detectors.py``) because c_impl mirror is dtd.h/dtd.c
— matching file boundaries simplifies Python ↔ C cross-reference.

Self-contained: depends only on numpy.
"""
import numpy as np


class DtdEstimator:
    """Double-Talk Detector with per-mode strategy.

    - 'geigel' mode (LMS/NLMS): Geigel DTD with hangover + confidence
    - 'divergence' mode (frequency-domain modes): Output-vs-input divergence detection
    - 'coherence' mode (frequency-domain modes): Error-reference coherence DT detection
    """

    def __init__(self, mode: str = 'geigel', *,
                 window_blocks: int = 4,
                 geigel_threshold: float = 0.5,
                 hangover_max: int = 15,
                 divergence_factor: float = 1.5,
                 attack: float = 0.3,
                 release: float = 0.05,
                 warmup_frames: int = 50,
                 # Coherence mode params
                 n_freqs: int = 0,
                 coh_alpha: float = 0.85,
                 coh_high: float = 0.6,
                 coh_low: float = 0.3,
                 coh_energy_floor: float = 0.01,
                 coh_abs_floor: float = 1e-6,
                 sample_rate: int = 16000,
                 block_size: int = 512):
        self.mode = mode  # 'geigel', 'divergence', or 'coherence'
        self.confidence = 0.0
        self.attack = attack
        self.release = release
        self.warmup_frames = warmup_frames
        self.frame_count = 0

        # Geigel state
        self.far_abs_buffer = np.zeros(max(1, window_blocks))
        self.buf_idx = 0
        self.geigel_threshold = geigel_threshold
        self.hangover_max = hangover_max
        self.hangover_count = 0

        # Divergence state
        self.divergence_factor = divergence_factor

        # Coherence state
        self.coh_alpha = coh_alpha
        self.coh_high = coh_high
        self.coh_low = coh_low
        self.coh_energy_floor = coh_energy_floor
        self.coh_abs_floor = coh_abs_floor  # #8: Absolute energy floor
        if mode == 'coherence' and n_freqs > 0:
            self.S_ex = np.zeros(n_freqs, dtype=np.complex64)
            self.S_ee = np.zeros(n_freqs, dtype=np.float32)
            self.S_xx = np.zeros(n_freqs, dtype=np.float32)
            # #7: Voice-band weighting (300Hz-4kHz emphasized)
            self.voice_weight = np.ones(n_freqs, dtype=np.float32)
            freq_per_bin = sample_rate / block_size
            for k in range(n_freqs):
                f = k * freq_per_bin
                if 300.0 <= f <= 4000.0:
                    self.voice_weight[k] = 3.0  # 3× weight for speech band
                elif f < 100.0 or f > 6000.0:
                    self.voice_weight[k] = 0.3  # De-weight extremes
        else:
            self.S_ex = None
            self.S_ee = None
            self.S_xx = None
            self.voice_weight = None

    def reset(self):
        self.confidence = 0.0
        self.frame_count = 0
        self.far_abs_buffer.fill(0)
        self.buf_idx = 0
        self.hangover_count = 0
        if self.S_ex is not None:
            self.S_ex.fill(0)
            self.S_ee.fill(0)
            self.S_xx.fill(0)

    def _update_confidence(self, detected: bool):
        """Update confidence with attack/release + hangover."""
        if detected:
            self.hangover_count = self.hangover_max
            self.confidence = min(self.confidence + self.attack, 1.0)
        elif self.hangover_count > 0:
            self.hangover_count -= 1
            self.confidence = max(self.confidence - self.release * 0.5, 0.0)
        else:
            self.confidence = max(self.confidence - self.release, 0.0)

    def _detect_geigel(self, near_end: np.ndarray, far_end: np.ndarray):
        """Geigel DTD: |mic| > threshold × max(|ref|) over window."""
        # Update far-end max circular buffer
        self.far_abs_buffer[self.buf_idx] = np.max(np.abs(far_end))
        self.buf_idx = (self.buf_idx + 1) % len(self.far_abs_buffer)
        far_max = np.max(self.far_abs_buffer)

        # Geigel test
        near_max = np.max(np.abs(near_end))
        detected = (far_max > 1e-6) and (near_max > self.geigel_threshold * far_max)

        self._update_confidence(detected)

    def _detect_divergence(self, near_end: np.ndarray, output: np.ndarray):
        """Output-vs-input divergence detection (output > input).

        Uses both energy-based and peak-based detection. Peak-based catches
        localized spikes that energy-based misses (e.g., transition transients).
        """
        output_energy = np.mean(output ** 2)
        near_energy = np.mean(near_end ** 2)
        output_peak = np.max(np.abs(output))
        near_peak = np.max(np.abs(near_end))

        if near_energy < 1e-10 and near_peak < 1e-6:
            # Silence → release
            self.confidence = max(self.confidence - self.release, 0.0)
            return

        # Check both energy and peak divergence
        energy_ratio = output_energy / (near_energy + 1e-10) if near_energy > 1e-10 else 0.0
        peak_ratio = output_peak / (near_peak + 1e-10) if near_peak > 1e-6 else 0.0
        ratio = max(energy_ratio, peak_ratio)

        mild_threshold = 1.2  # ratio < 1.2 is normal (unconverged, not diverging)
        if ratio > self.divergence_factor:
            # Severe divergence
            self.confidence = min(self.confidence + self.attack, 1.0)
        elif ratio > mild_threshold:
            # Mild divergence — proportional attack
            self.confidence = min(
                self.confidence + self.attack * (ratio - mild_threshold), 1.0)
        else:
            # Normal — faster release when ratio is well below 1.0
            release_scale = max(1.0 - ratio, 0.2)  # 0.2x ~ 1.0x
            self.confidence = max(
                self.confidence - self.release * (1.0 + 4.0 * release_scale), 0.0)

    def _detect_coherence(self, error_spec: np.ndarray, far_spec: np.ndarray):
        """Coherence-based double-talk detection.

        Uses smoothed magnitude-squared coherence between error and far-end.
        Low coherence + high error energy → near-end speech present → DT.
        High coherence → residual echo (unconverged) → keep updating.
        """
        alpha = self.coh_alpha

        # Update smoothed PSDs
        cross = error_spec * np.conj(far_spec)
        self.S_ex = alpha * self.S_ex + (1 - alpha) * cross
        self.S_ee = alpha * self.S_ee + (1 - alpha) * np.abs(error_spec) ** 2
        self.S_xx = alpha * self.S_xx + (1 - alpha) * np.abs(far_spec) ** 2

        # #7: Voice-band weighted coherence (ratio-of-sums)
        w = self.voice_weight
        num = np.sum(w * np.abs(self.S_ex) ** 2)
        den = np.sum(w * self.S_ee * self.S_xx)
        coherence = num / (den + 1e-10)

        # Energy check: only declare DT if error has meaningful energy
        sum_ee = np.sum(self.S_ee)
        sum_xx = np.sum(self.S_xx)
        # #8: Absolute energy floor prevents false triggers on quiet far-end
        has_energy = (sum_ee > self.coh_energy_floor * sum_xx and
                      sum_xx > 1e-10 and
                      sum_ee > self.coh_abs_floor)

        if coherence > self.coh_high:
            # Correlated → residual echo, not DT → release
            self._update_confidence(False)
        elif coherence < self.coh_low and has_energy:
            # Uncorrelated + energy → near-end speech → DT
            self._update_confidence(True)
        else:
            # Ambiguous → slow release
            self.confidence = max(self.confidence - self.release * 0.5, 0.0)

    def detect_block(self, near_end: np.ndarray, far_end: np.ndarray,
                     output: np.ndarray = None,
                     error_spec: np.ndarray = None,
                     far_spec: np.ndarray = None) -> float:
        """Update DTD state and return confidence [0.0, 1.0].

        For geigel mode: uses near_end and far_end.
        For divergence mode: uses near_end and output.
        For coherence mode: uses error_spec and far_spec.
        """
        self.frame_count += 1
        # Warmup: all detectors share the same warmup period.
        # Coherence also needs warmup because unconverged filter → error ≈ echo
        # → coherence estimate is unreliable (false DT triggers).
        if self.frame_count < self.warmup_frames:
            return 0.0

        if self.mode == 'geigel':
            self._detect_geigel(near_end, far_end)
        elif self.mode == 'coherence':
            if error_spec is not None and far_spec is not None:
                self._detect_coherence(error_spec, far_spec)
        else:
            self._detect_divergence(near_end, output)

        return self.confidence
